src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+
+            let (is_exit, time) = if let (Some(is_exit), Some(time)) = (parts.next(), parts.next()) {
+                let is_exit = match is_exit {
+                    "1" => true,
+                    "0" => false,
+                    a => panic!(format!("uh oh: {}", a)),
+                };
+
+                let time = time.parse::<f32>().unwrap();
+
+                (is_exit, time)
+            } else {
+                continue;
+            };
+
+            if is_exit && current_stack.is_empty() {
+                warn!("Found function exit without corresponding entrance. Discarding line. Check your input.");
+                continue;
+            }
+
+            {
+                let current = current_stack.current();
+                let duration = SCALE_FACTOR * (time - prev_start_time);
+
+                // hash `current` for lookup and insertion.
+                let mut hasher = stacks.hasher().build_hasher();
+                current.hash(&mut hasher);
+                let hash = hasher.finish();
+
+                match stacks
+                    .raw_entry_mut()
+                    .from_key_hashed_nocheck(hash, current)
+                {
+                    RawEntryMut::Occupied(mut occ) => *occ.get_mut() += duration,
+                    RawEntryMut::Vacant(vacant) => {
+                        // `Box<str>` has same hash as `str`.
+                        vacant.insert_hashed_nocheck(hash, Box::from(current), duration);
+                    }
+                }
+            }
+
+            if is_exit {
+                current_stack.pop();
+            } else {
+                let _ = parts.next();
+                let func_name = parts.next();
+                let _ = parts.next();
+                let path_name = parts.next();
+
+                if let (Some(func_name), Some(path_name)) = (func_name, path_name) {
+                    current_stack.call(func_name, path_name);
+                }
+            }
+
+            prev_start_time = time;
+        }
+
+        for (key, value) in stacks {
+            writeln!(writer, "{} {}", current_stack.frames(&key), value)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CallStack {
+    /// Create a function name interning call stack tracker.
+    ///
+    /// Popoulated with the constant builtins for inclusion, to enable a faster comparison.
+    pub fn new() -> Self {
+        CallStack {
+            strings: CALLS
+                .iter()
+                .enumerate()
+                .map(|(idx, name)| (name.to_owned().into(), idx))
+                .collect(),
+            interned_string: CALLS.iter().cloned().map(Rc::from).collect(),
+            calls: HashMap::new(),
+            interned: Vec::new(),
+            stack: Vec::with_capacity(16),
+        }
+    }
+
+    /// Push a `call` line on top of the stack.
+    pub fn call(&mut self, name: &str, path: &str) {
+        let new_or_not = match self.intern_str(name) {
```

Comment:
I can't quite wrap my head around this code. A comment explaining what it does and why would be helpful.